src/Gateway.php

```diff
@@ -0,0 +1,490 @@
+
+			exit;
+		}
+
+		/**
+		 * Step 1: Create a payment session
+		 * 
+		 * @link https://docs.adyen.com/online-payments/web-drop-in#create-payment-session
+		 */
+		$request = new PaymentSessionRequest(
+			AmountTransformer::transform( $payment->get_total_amount() ),
+			$this->config->get_merchant_account(),
+			$payment_id,
+			$this->get_payment_return_url( $payment_id )
+		);
+
+		// Complement payment request.
+		PaymentRequestHelper::complement( $payment, $request, $this->config );
+
+		// Payment method.
+		$payment_method = $payment->get_payment_method();
+
+		if ( null !== $payment_method ) {
+			// Payment method type.
+			$payment_method_type = PaymentMethodType::transform( $payment_method );
+
+			if ( null !== $payment_method_type ) {
+				$request->set_allowed_payment_methods( [ $payment_method_type ] );
+			}
+		}
+
+		try {
+			$payment_session = $this->client->create_payment_session( $request );
+		} catch ( \Exception $e ) {
+			Plugin::render_exception( $e );
+
+			exit;
+		}
+
+		// Register scripts.
+		$url_script = sprintf(
+			'https://checkoutshopper-%s.adyen.com/checkoutshopper/sdk/%s/adyen.js',
+			( self::MODE_TEST === $payment->get_mode() ? 'test' : 'live' ),
+			self::SDK_VERSION
+		);
+
+		// phpcs:ignore WordPress.WP.EnqueuedResourceParameters.MissingVersion -- Version is part of URL.
+		wp_register_script(
+			'pronamic-pay-adyen-checkout',
+			$url_script,
+			[],
+			null,
+			false
+		);
+
+		$file = '../js/dist/checkout-drop-in.js';
+
+		\wp_register_script(
+			'pronamic-pay-adyen-checkout-drop-in',
+			\plugins_url( $file, __FILE__ ),
+			[ 'pronamic-pay-adyen-checkout' ],
+			\hash_file( 'crc32b', __DIR__ . '/' . $file ),
+			true
+		);
+
+		// Register styles.
+		$url_stylesheet = sprintf(
+			'https://checkoutshopper-%s.adyen.com/checkoutshopper/sdk/%s/adyen.css',
+			( self::MODE_TEST === $payment->get_mode() ? 'test' : 'live' ),
+			self::SDK_VERSION
+		);
+
+		// phpcs:ignore WordPress.WP.EnqueuedResourceParameters.MissingVersion -- Version is part of URL.
+		wp_register_style(
+			'pronamic-pay-adyen-checkout',
+			$url_stylesheet,
+			[],
+			null
+		);
+
+		/**
+		 * Adyen checkout configuration.
+		 *
+		 * @link https://docs.adyen.com/checkout/drop-in-web
+		 * @link https://docs.adyen.com/checkout/components-web
+		 */
+		$configuration = [
+			'locale'                      => Util::get_payment_locale( $payment ),
+			'environment'                 => ( self::MODE_TEST === $payment->get_mode() ? 'test' : 'live' ),
+			'session'                     => (object) [
+				'id'          => $payment_session->get_id(),
+				'sessionData' => $payment_session->get_data(),
+			],
+			'clientKey'                   => $this->config->client_key,
+			'amount'                      => AmountTransformer::transform( $payment->get_total_amount() )->get_json(),
+			'paymentMethodsConfiguration' => $this->get_payment_methods_configuration( $payment ),
+		];
+
+		$configuration = (object) $configuration;
```

Comment:
Re-add auto submit functionality in `checkout-drop-in.js` and set  `$configuration['showPayButton'] = false;` here if only a single payment method from the below listed methods will be shown in drop-in.

```php
$auto_submit_methods = array(
	PaymentMethodType::SWISH,
	PaymentMethodType::TWINT,
	PaymentMethodType::VIPPS,
	PaymentMethodType::UNIONPAY,
);
```

Previously in `checkout-drop-in.js`:
https://github.com/pronamic/wp-pronamic-pay-adyen/blob/f48c76409c12b29927f5974d12d0c98361bbb9bf/js/src/checkout-drop-in.js#L130-L136

This avoids customers needing to select the payment method twice (e.g. in WooCommerce checkout and one more time in drop-in, even though there are no additional input fields for these methods in drop-in).

Reply:
How do we know for sure that the dropin doesn't want to show additional input fields for these methods? Maybe no additional fields are shown now, but by using the dropin this can change? If no additional input is needed, can we use a direct payment flow via the `/payments` endpint?

Reply:
As discussed at Pronamic HQ, we could keep the pay button and try to submit drop-in for these methods. If additional input is needed in the future, this should then result in input validation errors (before the transaction is started at Adyen). Previously, we were also hiding the pay button to prevent duplicate transactions for a single payment, but that is handled differently now with the sessions implementation.

**Swish**
Country Code: `SE`
Currency: `SEK`

The Swish payment flow uses a QR code (desktop) or a redirect to the Swish app (redirect URL with `swish://` protocol). Drop-in can handle this for us.

FYI, if I just start a Swish payment via the API integration, the following error occurs:

> The handling of the `Pending` result code is not implemented.

**TWINT**
Country Code: `CH`
Currency: `CHF`

Seems to work with our API integration, with `RedirectShopper` result code.

**Vipps**
Country Code: `NO`
Currency: `NOK`

Seems to work with our API integration, with `RedirectShopper` result code.

**China UnionPay**
I can't remember how this one ended up in the auto-submit list. Seems to work fine with drop-in, without additional input fields. However, with our API integration, I get the following error message:

> 交易失败 [9200003]Empty req.

I'd suggest just using drop-in for this payment method.

Reply:
For documentation purposes I created a separate issue for this:

- https://github.com/pronamic/wp-pronamic-pay-adyen/issues/9

Reply:
https://github.com/pronamic/wp-pronamic-pay-adyen/issues/9#issuecomment-1128829915